Improved autosaves

krs0 opened the pull request:
First working version of this proposal: #2561

Code will be rearranged, some placement suggestions (for helper functions in special) would be welcomed.

There are some hardcoded things still.

- [x] I do not know how to find out if game is hotseat

- [x] I do not know how to change from startTimeIso8601 to a more readable format: 20190414T111005 vs 2019-04-14_11-10-05 (there is a class variable for startTimeIso8601 should I just create another one)?
replaced the startTimeIso8601 variable with a more human readable one.

- [x] map name templates should move to launcher as 4 separate settings.
settings created, will not expose them in GUI

If someone could take a look would be great.

Laserlicht commented on client/CPlayerInterface.cpp:
```diff
@@ -198,48 +199,200 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
+		// save using a counter, and overwrite depending on autosave count limit
+		if(autosaveCountLimit > 0) 
+		{
+			auto saveNumber = extractNumberAfterHash(lastMapSaveName);
```
Wouldn't it better and cleaner to save this information in the save file itself and get it from there? Then you can also make the counter as variable.

krs0 replied:
will look how to make that.

krs0 replied:
I think save file should not store this information. For now this will work.
Maybe in the future we get rid of it altogether? Or change how it works if Quick Load / Quick Save is implemented?

IvanSavenko commented on client/CPlayerInterface.cpp:
```diff
@@ -198,48 +199,184 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
 		}
-	}
+
+		saveName = replaceVariables(saveName, variables);
```
You can use `boost::replace_all` instead of this method, e.g.
```cpp
for (auto const & entry : variables)
    boost::replace_all(saveName, "%" + entry.first + "%", entry.second);
```

krs0 replied:
is boost not shunned everywhere in favor of std:: ?

IvanSavenko replied:
If there is *identical* functionality in std - maybe.
But here boost provides way more convenient text formatting tools

krs0 replied:
Done

IvanSavenko commented on include/vstd/DateUtils.h:
```diff
@@ -4,10 +4,8 @@ VCMI_LIB_NAMESPACE_BEGIN
 	DLL_LINKAGE std::string getFormattedDateTime(std::time_t dt);
-	DLL_LINKAGE std::string getDateTimeISO8601Basic(std::time_t dt);
-
+	DLL_LINKAGE std::string getDateTimeFormatted(std::time_t dt);
```
What's the difference between `getFormattedDateTime` and `getDateTimeFormatted`?
At least one of them might need better name and probably comment to explain formatting rules

krs0 replied:
I have changed the names to:
getDateTimeLocalized()
getDateTimeFormatted() -> name suggests its a special format, does not give info on the format :( (I would keep it like this until another function arrives)

IvanSavenko commented on client/CPlayerInterface.cpp:
```diff
@@ -198,48 +199,184 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
+
+	// store variable values in map
+	std::unordered_map<std::string, std::string> variables;
+	variables["mapName"] = mapName;
```
You can add '%' in here and remove string manipulation during replacement:
```suggestion
	variables["%mapName%"] = mapName;

```

krs0 replied:
Done

krs0 replied:
done

krs0 commented:
Autosave limit makes sense only for simple saves or before battle saves.
When we save under the name: Autosave#1... 2,3,4,5, .... and then overwrite 1, then 2... etc.

In complex save naming, it makes no sense. Usually saves will be saved in named folders with turn info as filenames. Lime Broken Alliance/111, 112, 113..... so here limit makes no sense.

IvanSavenko commented:
If autosave limit does not makes sense in most cases - then do we need such option to begin with?

krs0 commented:
Is someone does not care about complex saving scheme, games will be saved like Autosave#1, 2, 3... etc
This option will come in handy to automatically limit the numbers of autosaves.
I suspect this scheme will be quite handy for android.

IvanSavenko commented:
And what is used by default?

My point is:
I ran game (current develop) with default settings and I keep getting autosaves in `Autosave/Arrogance_2023XXXXX/1.vsgm1` format.

So I presume that by default we use such complex naming scheme that (effectively) ignores autosave limit. Correct?

krs0 commented on client/CPlayerInterface.cpp:
```diff
@@ -198,48 +199,200 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
+		+ std::to_string(cb->getDate(Date::DAY_OF_WEEK)); 
+
+	// get random map info
+	if(mapName == "Random Map")
```
Good catch! ty. Replaced for now by mapGenOptions != nullptr; check.
Problem is the API related to maps campaigns and rmg maps is lacking a lot of functionality, and I do not trust myself to change those APIs.
I was expecting something straightforward like a bool property somwhere isRMGMap :(

dydzio0614 commented:
i have just one request: If this PR removes some autosave features, then I want to be able for autosave to work like in original H3 so I can have one AUTOSAVE file without folder

krs0 commented:
@dydzio0614 ok boomer :)

simple autosave - autosaves in SAVES/Autosave#, or SAVES/Autosave_xxx depending on autosave limit

Autosave limit = how many autosaves to keep.

krs0 commented:
Ok works as intended: (maybe I find better place for 1 or 2 variables, but functionality is complete)
If someone can look into changes I did to Launcher (I am not so sure on that editor :( )
Out of scope for this PR:
```
	// ToDo : krs - gather all static info only once at map start / load
	// ToDo : krs - autosaves before each battle
	// ToDo : krs - autosave at beginning of game
	// ToDo : krs - quick autosaves on hotkey (autosave mode Using Counter could be used)
```
Example folder hierarchy from my test run:
[Saves.zip](https://github.com/vcmi/vcmi/files/12579515/Saves.zip)

These are launcher settings:
![image](https://github.com/vcmi/vcmi/assets/15194321/79ee6fa8-50fa-470c-baa3-3e3375d92d8a)
![image](https://github.com/vcmi/vcmi/assets/15194321/04dc1a2d-5527-435f-b5af-f168f5c1c954)

krs0 commented:
Long long names are still an issue:
![image](https://github.com/vcmi/vcmi/assets/15194321/177c1caf-5f8f-4bf1-a052-c52adcb98011)

krs0 commented:
Ok I have moved some code out of performAutosave and added a limit in GUI to filenames and folder names.

If someone could review, would be great.

I have 2 functions I do not know where to put:

// Generic function to replace forbidden characters with underscores
std::string replaceForbiddenChars(const std::string& input)

// Generic function to concatenate a vector of strings with a specified delimiter
std::string concatenateWithDelimiter(const std::vector<std::string>& strings, const std::string& delimiter)

krs0 commented:
I intend to try! to clean up a little all the places were there is map info (or where these should be :( ). But not in this PR.

IvanSavenko commented on client/CPlayerInterface.cpp:
```diff
@@ -211,47 +212,189 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
 }
 
-void CPlayerInterface::performAutosave()
+int extractNumberAfterHash(const std::string& input) 
```
1) Either make this function part of PlayerInterface class or mark it as 'static'
2) Perhaps move all autosave-related logic outside of player interface into a separate class? PlayerInterface already has such similar extracted components

krs0 replied:
extractNumberAfterHash has nothing specific to PlayerInterface so I moved it to TextOperations along with the other string operating functions

IvanSavenko commented on lib/StartInfo.cpp:
```diff
@@ -37,13 +37,47 @@ bool PlayerSettings::isControlledByHuman() const
+	return playerNames;
+}
+
+uint8_t StartInfo::getNoOfHumanPlayers() const
```
```suggestion
uint8_t StartInfo::humanPlayerCount() const
```

krs0 replied:
uint8_t getHumanPlayersCount() const;

krs0 commented:
This will be unmergeable once Launcher is updated. If no other suggestions could this be merged?
Improvements like below are independent, I will list them in the original issue:
```
	// ToDo : krs - gather all static info only once at map start / load
	// ToDo : krs - autosaves before each battle
	// ToDo : krs - autosave at beginning of game
	// ToDo : krs - quick autosaves on hotkey (autosave mode Using Counter could be used)
```

IvanSavenko commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -40,6 +40,14 @@ static const std::string cursorTypesList[] =
 	"software"
 };
 
+static const std::string autosaveModeList[] =
+{
+	"Off",
+	"Using Counter",
+	"Using Turn Info",
+	"Using Schema"
+};
+
```
This strings should be added directly into widget to make them translateable

krs0 replied:
using now custom data for controls:

```	ui->comboBoxAutosaveMode->setItemData(0, "Off", Qt::UserRole);
	ui->comboBoxAutosaveMode->setItemData(1, "Using Counter", Qt::UserRole);
	ui->comboBoxAutosaveMode->setItemData(2, "Using Turn Info", Qt::UserRole);
	ui->comboBoxAutosaveMode->setItemData(3, "Using Schema", Qt::UserRole);
```

krs0 commented on lib/StartInfo.cpp:
```diff
@@ -37,13 +37,47 @@ bool PlayerSettings::isControlledByHuman() const
+		if(!playerInfos.count(player))
+			continue;
+
+		auto playerInfo = playerInfos.at(player);
+		if(playerInfos.at(player).isControlledByHuman())
```
Done

krs0 commented on client/CPlayerInterface.cpp:
```diff
@@ -224,47 +225,135 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
+		Campaign,
+		Scenario,
+		RandomMap,
+		Undefined = 255
```
Done

krs0 commented on client/CPlayerInterface.cpp:
```diff
@@ -224,47 +225,135 @@ void CPlayerInterface::playerStartsTurn(PlayerColor player)
 }
 
-void CPlayerInterface::performAutosave()
+void CPlayerInterface::performAutosave() 
```
Latest merging was time consuming. If really needed could this be done in separate PR?

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -210,4 +248,28 @@ std::string TextOperations::escapeString(std::string input)
 
+int TextOperations::extractNumberAfterHash(const std::string & input)
+{
+    size_t hashPos = input.find('#');
```
Use tabs instead of spaces

krs0 replied:
Done. I made copy past mistake, that was not caught because my VS automatic tools are off.
If not everybody sticks with code formatting guide, VS will change everything and then will let you wonder what are changes and what is whitespaces.

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -210,4 +248,28 @@ std::string TextOperations::escapeString(std::string input)
+{
+    size_t hashPos = input.find('#');
+	if(hashPos != std::string::npos)
+		return -1;
```
So function extracts only non-negative number, not a number

krs0 replied:
/// extracts the number after a # character (useful in autosave filenames)

I think should be obvious without renaming the function. If any suggestion I can rename.

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -210,4 +248,28 @@ std::string TextOperations::escapeString(std::string input)
+	if(hashPos != std::string::npos)
+		return -1;
+
+	std::string numberAsString = input.substr(hashPos + 1); // Get the substring after '#'
```
Only one digit??

krs0 replied:
Bug!
`std::string numberAsString = input.substr(hashPos + 1, input.length() - (hashPos + 1)); // Get the substring after '#' until the end of the string`

krs0 replied:
Bug!
`std::string numberAsString = input.substr(hashPos + 1, input.length() - (hashPos + 1)); // Get the substring after '#' until the end of the string`

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -199,6 +199,44 @@ size_t TextOperations::getUnicodeCharactersCount(const std::string & text)
+    return result;
+}
+
+std::string TextOperations::replaceForbiddenPathChars(const std::string & input)
```
Suggest to add delimiter as separate argument, instead of hardcoded underscore

krs0 replied:
added delimiter as parameter

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -199,6 +199,44 @@ size_t TextOperations::getUnicodeCharactersCount(const std::string & text)
+    std::string forbiddenChars = "\\/<>:\"|?*\t\n\v\f\r"; // space is not here since folder and file names can have spaces!
+    std::string result = input;
+
+    for (char& c : result)
+    {
+        if(forbiddenChars.find(c) != std::string::npos)
+            c = '_';
+    }
```
```suggestion
for(char c : forbiddenChars)
std::replace(result.begin(), result.end(), c, delimiter);
```

krs0 replied:
Done

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -199,6 +199,44 @@ size_t TextOperations::getUnicodeCharactersCount(const std::string & text)
+{
+    std::string result;
+    
+    for(const std::string& str : strings) {
+        result += str + delimiter;
+    }
+
+    if(!result.empty() && result.size() >= delimiter.size()) {
+        result.erase(result.end() - delimiter.size(), result.end());
+    }
+
+    return result;
```
Something strange: why not iterating over iterators and skip adding delimiter for the last one?

krs0 replied:
changed as suggested.

Nordsoft91 commented on lib/TextOperations.cpp:
```diff
@@ -199,6 +199,44 @@ size_t TextOperations::getUnicodeCharactersCount(const std::string & text)
 	return conv.from_bytes(text).size(); 
 }
 
+std::string TextOperations::shortenUnicodeString(const std::string & input, int maxLength)
```
More common name is truncate/truncateUnicodeString

krs0 replied:
Done

Nordsoft91 commented on lib/StartInfo.cpp:
```diff
@@ -58,13 +58,47 @@ bool PlayerSettings::isControlledByHuman() const
+
+uint8_t StartInfo::getHumanPlayersCount() const
+{
+	uint8_t NoOfHumanPlayers = 0;
```
Local variables start with non-capital in vcmi code

krs0 replied:
renamed to humanPlayersCount

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -540,3 +593,17 @@ void CSettingsView::on_spinBoxReservedArea_valueChanged(int arg1)
+	ui->comboBoxAutosaveMode->addItem("Using Turn Info");
+	ui->comboBoxAutosaveMode->setItemData(2, "Using Turn Info", Qt::UserRole);
+
+	ui->comboBoxAutosaveMode->addItem("Using Schema");
```
Using schema -> using pattern

krs0 replied:
Renamed to Using Template. Seems more understandable?

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -515,23 +569,22 @@ void CSettingsView::on_comboBoxAlliedPlayerAI_currentTextChanged(const QString &
-    node->Bool() = arg1;
-    ui->lineEditAutoSavePrefix->setEnabled(arg1);
+    Settings node = settings.write["general"]["autosaveFrequency"];
+    node->Float() = arg1;
```
Why float not integer?

krs0 replied:
because other controls use the same. Copy paste -> changed to Integet

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -315,10 +345,34 @@ void CSettingsView::on_comboBoxShowIntro_currentIndexChanged(int index)
-	Settings node = settings.write["general"]["saveFrequency"];
-	node->Integer() = index;
+	QVariant itemData = ui->comboBoxAutosaveMode->itemData(index, Qt::UserRole);
+	std::string selectedEntry = itemData.toString().toStdString();
```
Please, don't store such strings in item data. Instead, create enum and store enum value in item data and them compare with enum instead of string comparison

krs0 replied:
I have started with Enum but Ivan pointed out that there could be problems with translation. I did not understand the remark but in lack of feedback, a quick search on the net showed this itemData as best approach?

I see no problems with it, but if needed  I could revert to previous version that used enums.

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -40,6 +40,26 @@ static const std::string cursorTypesList[] =
 
+int findComboBoxItemByInternalName(QComboBox* comboBox, const std::string& targetInternalName)
+{
+    int itemCount = comboBox->count();
+    int foundIndex = -1;
+
+    for (int index = 0; index < itemCount; ++index)
+    {
+        QVariant itemData = comboBox->itemData(index, Qt::UserRole);
+        if (itemData.toString().toStdString() == targetInternalName)
+        {
+            foundIndex = index;
+            break;
+        }
+    }
+
+	assert(foundIndex != -1);
+
+    return foundIndex;
```
```suggestion
    for (int index = 0; index < comboBox->count(); ++index)
    {
        QVariant itemData = comboBox->itemData(index, Qt::UserRole);
        if (itemData.toString().toStdString() == targetInternalName)
            return index;
    }

	assert(0);

    return -1;
```

krs0 replied:
changed as suggested

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -40,6 +40,26 @@ static const std::string cursorTypesList[] =
 	"software"
 };
 
+int findComboBoxItemByInternalName(QComboBox* comboBox, const std::string& targetInternalName)
```
It doesn't find item, but index

krs0 replied:
renamed to findComboBoxItemsIndexByInternalName

Nordsoft91 commented on client/lobby/SelectionTab.cpp:
```diff
@@ -443,13 +448,17 @@ void SelectionTab::filter(int size, bool selectFirst)
 			std::shared_ptr<ElementInfo> folder = std::make_shared<ElementInfo>();
 			folder->isFolder = true;
 			folder->folderName = folderName;
+			folder->displayName = TextOperations::shortenUnicodeString(folder->folderName, folderNameMaxDisplayLength);
```
Don't you want to add ... in the end of string if truncated?

krs0 replied:
done

Nordsoft91 reviewed (changes requested):
Formatting is broken - spaces/tabs are messed. We use tabs in vcmi code

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.cpp:
```diff
@@ -540,3 +593,17 @@ void CSettingsView::on_spinBoxReservedArea_valueChanged(int arg1)
 
+void CSettingsView::addItemsToControls()
+{
+	ui->comboBoxAutosaveMode->addItem("Off");
```
Enable translations for those strings using tr()
```suggestion
	ui->comboBoxAutosaveMode->addItem(tr("Off"));
```

krs0 replied:
Done.

Nordsoft91 commented on launcher/settingsView/csettingsview_moc.h:
```diff
@@ -68,11 +68,11 @@ private slots:
-    void on_checkBoxAutoSavePrefix_stateChanged(int arg1);
+    void on_spinBoxAutosaveFrequency_valueChanged(int arg1);
 
-    void on_spinBoxAutoSaveLimit_valueChanged(int arg1);
+    void on_spinBoxAutosaveCountLimit_valueChanged(int arg1);
 
-    void on_lineEditAutoSavePrefix_textEdited(const QString &arg1);
+    void on_checkBoxAutosaveBeforeBattles_stateChanged(int arg1);
```
some legacy formatting with spaces

krs0 commented:
Again conflicts on merging :( Will not maintain it further.

krs0 commented:
Requirements are not matching VCMI vision. Dropping this PR.
